File: src/com/company/java_examples/GreatestCommonDivisor.java
package com.company.java_examples;

public class GreatestCommonDivisor {

    public static void main(String[] args) {
        int num1 = 12;
        int num2 = 16;

        int result = greatestCommonDivisor(num1, num2);
        System.out.println(result);
    }

    private static int greatestCommonDivisor(int num1, int num2) {
        // a = q * b + r
        // then gcd(a, b) = gcd(b, r)

        // loop till remainder is 0
        while (num2 > 0) {
            int reminder = num1 % num2; // remainder
            num1 = num2;
            num2 = reminder;
        }
        return num1;

    }

    private static int greatestCommonDivisorRecursive(int num1, int num2) {
        // a = x * b + c
        // then gcd(a, b) = gcd(b, c)

        if (num2 == 0) {
            return num1;
        }
        return greatestCommonDivisorRecursive(num2, num1 % num2);
    }
}

File: src/com/company/kotlin_examples/GreatestCommonDivisor.kt
package com.company.kotlin_examples

object GreatestCommonDivisor {

    @JvmStatic
    fun main(args: Array<String>) {
        val num1 = 12
        val num2 = 16
        val result = greatestCommonDivisor(num1, num2)
        println(result)
    }

    private fun greatestCommonDivisor(num1: Int, num2: Int): Int {
        // a = q * b + r
        // then gcd(a, b) = gcd(b, r)

        // loop till remainder is 0
        var num1 = num1
        var num2 = num2
        while (num2 > 0) {
            val reminder = num1 % num2 // remainder
            num1 = num2
            num2 = reminder
        }
        return num1
    }

    private fun greatestCommonDivisorRecursive(num1: Int, num2: Int): Int {
        // a = x * b + c
        // then gcd(a, b) = gcd(b, c)
        return if (num2 == 0) {
            num1
        } else greatestCommonDivisorRecursive(num2, num1 % num2)
    }
}